### PR TITLE
Improved Area Light Support for Hair

### PR DESCRIPTION
**Pull request by johnpars**
### Purpose of this PR
This PR adds an improved method for supporting rectangular area lights for hair. 

Due to lack of support for LTC with anisotropic models, we have up until now fallen back to the GGX LTC for hair. This PR tries to improve on this by instead deferring to a "Representative Point" as a stop-gap method for both the Approximate (Kajiya) and Physical (Marschner) based model until further research is given for LTC. 

The result of this work is that we should now be much closer to our references. 

Below is some results for our approximation for Physical.

image

And here is some results for our approximation for Approximate. 
[TODO IMAGE]

---
### Performance

Performance-wise this seems like a slight win over LTC for Marschner model and a pretty significant win for Kajiya. Additionally, I measured the MRP method with and without the brute force solid angle computation and did find a measurable difference, but in both cases these still resulted in lower times than the LTC, so I think this is good justification to stick with the brute force solid angle route. 

Performance metrics were captured from a standalone Windows built with an NVIDIA Quadro RTX 8000, at a 3840x2160 resolution, 8x MSAA. 

| Model | Area Light Method  | Forward Opaque Time Per-Groom (ms) |
| -------------| ------------- | ------------- |
| Marschner (+Multiple Scattering) | MRP, Brute Force Solid Angle | 1.31 |
| Kajiya-Kay | MRP, Brute Force Solid Angle | 0.97  |
| Marschner (+Multiple Scattering) | MRP, Optimized Solid Angle | 1.27 |
| Kajiya-Kay | MRP, Optimized Solid Angle | 0.89 |
| Marschner (GGX) | LTC | 1.57 |
| Kajiya-Kay (GGX) | LTC | 1.43  |

_Test scenario scene. Left Kajiya, right Marschner._ 
image

---
### Testing status

I manually verified the following behaviors:
- Varying area light shapes and orientations are in good agreement with the reference. 
- Cookie support.
- Barn door support. 
- Shadow support. 

Updated the hair area light graphics test to reflect the improved area light support, as well as added more coverage for the marschner model.  Local tests are passing + launched yamato.

1401_HairGraph_Area_Light

---
### Comments to reviewers
I ended up finding a reasonable way to approximate Kajiya with this approach. It doesn't exactly use MRP, but instead just treats the area light as a normal analytic light and modulates the roughness automatically based on the solid angle. It required some tuning to make sure we are in good agreement with the reference.

### CODE REVIEW

File: com.unity.render-pipelines.high-definition/Runtime/Material/Hair/Hair.hlsl [hunk 1]
```diff
@@ -1052,6 +1134,152 @@ DirectLighting EvaluateBSDF_Rect(   LightLoopContext lightLoopContext,
...
+
+        // Only apply cookie if there is one
+        if ( lightData.cookieMode != COOKIEMODE_NONE )
+        {
+            // Compute cookie's mip count.
+            const float cookieWidth = lightData.cookieScaleOffset.x * _CookieAtlasSize.x; // Guaranteed power of two.
+            const float cookieMips  = round(log2(cookieWidth));
+
+            // Get the 4x4 mip level.
+            const float cookieMip = cookieMips - 2;
```
[the remark this answers is not in the document]
Comment (johnpars) [reply com.unity.render-pipelines.high-definition/Runtime/Material/Hair/Hair.hlsl]: > the mip level should be sensitive to the solid angle. Larger solid angles indicate that the light is closer (relative to its size), so a smaller portion of its surface contributes to shading, and we should pick a larger mip.

I am getting good results with this mip selection approach:
`MipLevel = MipCount - (MipCount * SolidAngle / 2PI)`


For roughness, I tested the reference and it didn't seem worth it to take it into account. 

SolidAngleMip

File: com.unity.render-pipelines.high-definition/Runtime/Material/Hair/MultipleScattering/HairMultipleScattering.hlsl [hunk 2]
```diff
@@ -164,7 +164,7 @@ float3 EvaluateMultipleScattering(float3 L, float3 Fs, BSDFData bsdfData, float3
 
     // Resolve the overall local scattering term ( Eq. 19 & 20 Disney ).
     // Note, for now remove the square, there seems to be a discrepancy in the gaussian used in the paper (ours takes std. dev)
-    float3 fsBack = db * 2 * Ab * D_LongitudinalScatteringGaussian(thetaH - deltaB, sqrt(sigmaB + sigmaF)) / PI;
+    float3 fsBack = db * 2 * Ab * D_LongitudinalScatteringGaussian(thetaH - deltaB, sqrt(sigmaB + sigmaF)) / 1;
```
[the remark this answers is not in the document]
Comment (kecho) [reply com.unity.render-pipelines.high-definition/Runtime/Material/Hair/MultipleScattering/HairMultipleScattering.hlsl:167]: Could you ensure this doesn't change tests? i.e. do some yamato runs and update refs if needed.
Also I would add this comment in the code itself as  a comment! that is all :)

**Review by kecho — CHANGES_REQUESTED** [review]
Run yamato, and add a comment on the removal of that division by PI :) ensure yamato is green!

**Comment by johnpars** [comment]
Re-running Yamato. All of these failures seem like instabilities (including the smoke test one, still persisting after I merged master). I also provided QA a test project to validate this PR.

**Review by robertcupisz — APPROVED** [review]

**Review by kecho — APPROVED** [review]

**Comment by johnpars** [comment]
Looks like we are 'green' test-wise, the failures are known instabilities.

File: com.unity.render-pipelines.high-definition/Runtime/Material/Hair/Hair.hlsl [hunk 3]
```diff
@@ -816,20 +823,97 @@ CBSDF EvaluateBSDF(float3 V, float3 L, PreLightData preLightData, BSDFData bsdfD
...
 
-// Secondary shadow tap that can provide a higher quality occlusion information for the multiple scattering.
-#define LIGHT_EVALUATION_SPLINE_SHADOW_VISIBILITY_SAMPLE
 #endif
 
 #include "Packages/com.unity.render-pipelines.high-definition/Runtime/Lighting/LightEvaluation.hlsl"
 #include "Packages/com.unity.render-pipelines.high-definition/Runtime/Material/MaterialEvaluation.hlsl"
 #include "Packages/com.unity.render-pipelines.high-definition/Runtime/Lighting/SurfaceShading.hlsl"
 
+float3 RayPlaneIntersect(in float3 rayOrigin, in float3 rayDirection, in float3 planeOrigin, in float3 planeNormal)
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Material/Hair/Hair.hlsl:847]: Those set of function either alredy exist or should be moved into our geometry.hlsl common file

File: com.unity.render-pipelines.high-definition/Runtime/Material/Hair/Hair.hlsl [hunk 4]
```diff
@@ -1063,7 +1291,11 @@ DirectLighting EvaluateBSDF_Area(LightLoopContext lightLoopContext,
     }
     else
     {
+#if 0
```
[the remark this answers is not in the document]
Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/Runtime/Material/Hair/Hair.hlsl:1304]: We can't afford both in the code for perf reasons anyway (vgpr). So totally fine with disabling LTC

**Review by sebastienlagarde — APPROVED** [review]

**Comment by sebastienlagarde** [comment]
@QA: Merging to 22.1, the coverage by John is already great.

**Review by JMargevics — COMMENTED** [review]
Tested thoroughly this on Windows platforms and found no visual issues both with strands and hair cards. ✔👍

**Pull request merged**